### PR TITLE
Make it possible to bulk create and destroy excluded requests via API

### PR DESCRIPTION
**Pull request by eduardoj**
Co-authored-by: David Kang <dkang@suse.com>

### CODE REVIEW

**Comment by eduardoj** [comment]
[Exclude Request should be added with XML (API).](https://trello.com/c/lpKZnL9q/1010-exclude-request-should-be-added-with-xml-api)

File: src/api/db/structure.sql [hunk 1]
```diff
@@ -1140,6 +1140,7 @@ CREATE TABLE `staging_request_exclusions` (
   `description` varchar(255) COLLATE utf8mb4_unicode_ci DEFAULT NULL,
   `created_at` datetime NOT NULL,
   `updated_at` datetime NOT NULL,
+  `number` int(11) DEFAULT NULL,
```
Comment (ChrisBr) [inline src/api/db/structure.sql]: Why do we need to add the number now?

**Comment by ChrisBr** [comment]
I'm note sure if I understand but why is this PR named 

> Make it possible to the API to create and destroy excluded requests

as this is just a refactoring, no? What is the feature you added?

**Comment by eduardoj** [comment]
@ChrisBr:

> What is the feature you added?

Before excluded requests could only be created and destroyed in the API once at a time.

With this pull request xml like this will be accepted with the API for creating:

```
<excluded_requests>
  <request number='14' description='Hey!'/>
  <request number='15' description='What?'/>
</excluded_requests>
```
and like this for deleting excluded requests:
```
<requests>
  <number>14</number>
  <number>15</number>
</requests>
```

**Comment by ChrisBr** [comment]
@eduardoj thanks for the explanation, I changed the title to 

> Make it possible to bulk create and destroy excluded requests via API

to reflect that.

File: src/api/app/controllers/staging/excluded_requests_controller.rb [hunk 2]
```diff
@@ -1,54 +1,86 @@
...
     end
   end
 
   private
 
-  def set_staging_workflow
-    project = Project.get_by_name(params[:staging_main_project_name])
-    @staging_workflow = project.staging
-    return if @staging_workflow
+  def set_xml_hash
```
Comment (hennevogel) [inline src/api/app/controllers/staging/excluded_requests_controller.rb]: Bad naming, yes it's an `XmlHash` but what do you expect inside?

File: src/api/app/controllers/staging/excluded_requests_controller.rb [hunk 3]
```diff
@@ -1,54 +1,86 @@
...
 
-    if request_exclusion.save
-      render_ok
-    else
+    if errors?
       render_error(
         status: 400,
         errorcode: 'invalid_request',
-        message: request_exclusion.errors.full_messages.to_sentence
+        message: "Excluding requests to #{@staging_workflow} failed: #{errors_list.join('. ')}."
```
Comment (hennevogel) [inline src/api/app/controllers/staging/excluded_requests_controller.rb]: :thought_balloon:  s/to/for/

File: src/api/app/controllers/staging/excluded_requests_controller.rb [hunk 4]
```diff
@@ -1,54 +1,86 @@
...
+    [@xml_hash[:number]].flatten
   end
 
-  def set_request_exclusion
-    request = BsRequest.find_by!(number: params[:number])
-    @request_exclusion = request.request_exclusion
-    return if @request_exclusion
+  def requests_to_be_excluded
+    xml_hash_requests.map do |request|
+      bs_request = @staging_workflow.unassigned_requests.find_by_number(request[:number])
```
Comment (hennevogel) [inline src/api/app/controllers/staging/excluded_requests_controller.rb]: What if this is `nil`?

File: src/api/db/structure.sql (hunk 1 above)
Comment (eduardoj) [reply src/api/db/structure.sql]: @ChrisBr, we followed @coolo's [suggestion](https://github.com/openSUSE/open-build-service/pull/6201#discussion_r233365761). It was easier to add the 'number' column.

File: src/api/app/controllers/staging/excluded_requests_controller.rb (hunk 2 above)
Comment (eduardoj) [reply src/api/app/controllers/staging/excluded_requests_controller.rb]: @hennevogel:

> Bad naming, yes it's an `XmlHash` but what do you expect inside?

I expect inside to have an API request in XmlHash format.

Fine with changing `set_xml_hash` to `set_xml_hash_request`, and `@xml_hash` to `@xml_hash_request`?

File: src/api/app/controllers/staging/excluded_requests_controller.rb (hunk 4 above)
Comment (eduardoj) [reply src/api/app/controllers/staging/excluded_requests_controller.rb]: @hennevogel:

> What if this is `nil`?

This case is handled by validations. There is also a [test](https://github.com/openSUSE/open-build-service/pull/6774/files#diff-d6b6bb094c9652eb21cfdfa20e28ed77R58) for this case.

File: src/api/app/controllers/staging/excluded_requests_controller.rb (hunk 2 above)
Comment (hennevogel) [reply src/api/app/controllers/staging/excluded_requests_controller.rb]: yes, you expect a list of requests. requests_xml_hash or whatnot sounds fine to me...

Comment (eduardoj) [reply src/api/app/controllers/staging/excluded_requests_controller.rb]: Changed to requests_xml_hash.

**Comment by coolo** [comment]
I don't like this API at all (just as the XML to select and unselect). To construct this XML you need a lot of code - while in most cases you ignore single requests. So why insist on requiring XML as input?

**Comment by ChrisBr** [comment]
> I don't like this API at all (just as the XML to select and unselect). To construct this XML you need a lot of code - while in most cases you ignore single requests. So why insist on requiring XML as input?

So you would prefer like it was before that you can only exclude one request per api call? Same for select/unselect?

**Comment by coolo** [comment]
> > I don't like this API at all (just as the XML to select and unselect). To construct this XML you need a lot of code - while in most cases you ignore single requests. So why insist on requiring XML as input?
> 
> So you would prefer like it was before that you can only exclude one request per api call? Same for select/unselect?

Yes, especially as select and unselect are very lengthy processes (as you ditched the idea of making it background jobs) and as such I want to show progress on the command line.

**Comment by ChrisBr** [comment]
> Yes, especially as select and unselect are very lengthy processes (as you ditched the idea of making it background jobs) and as such I want to show progress on the command line.

Yo, background job is ditched FOR NOW but we still plan to do that in the (near) future. As this route is more flexible I would keep it, crafting the XML is anyway done by a bot / tool

**Pull request merged**